File: frontend/src/types/social.ts
export type UserStatus = 'online' | 'offline' | 'in-game' | 'away' | 'busy';

export interface SocialUser {
  id: string;
  username: string;
  avatar?: string;
  elo: number;
  status: UserStatus;
  currentActivity?: string;
  lastSeen?: string;
}

export interface Friend extends SocialUser {
  friendSince: string;
  isFavorite?: boolean;
  mutualFriends?: number;
}

export interface FriendRequest {
  id: string;
  fromUserId: string;
  fromUsername: string;
  fromUser: SocialUser;
  fromAvatar?: string;
  toUserId: string;
  status: 'pending' | 'accepted' | 'rejected';
  message?: string;
  createdAt: string;
}

export interface Message {
  id: string;
  fromUserId: string;
  fromUsername: string;
  toUserId: string;
  content: string;
  isRead: boolean;
  createdAt: string;
  // Chat-style fields — the data layer always provides these, so they're
  // required (no `?`). Keeps `Message.timestamp` / `.content` dereferences in
  // ChatInterface clean (no `?.` everywhere).
  conversationId: string;
  senderId: string;
  timestamp: string;
  status: 'sent' | 'delivered' | 'read';
  type: 'text' | 'image' | 'video' | 'system';
}

export interface Conversation {
  id: string;
  participantId: string;
  participantUsername: string;
  participantAvatar?: string;
  lastMessage?: Message;
  lastMessageAt?: string;
  unreadCount: number;
  // Chat-style fields — required because the data layer / UI both treat
  // these as always-present.
  type: 'direct' | 'group' | 'party';
  participants: SocialUser[];
  updatedAt: string;
  partyId?: string;
}

export interface Party {
  id: string;
  leaderId: string;
  leaderUsername: string;
  name: string;
  description?: string;
  maxMembers: number;
  currentMembers: number;
  members: PartyMember[];
  createdAt: string;
  // Party UX fields — data layer always supplies these.
  region: string;
  isPrivate: boolean;
  voiceChatEnabled: boolean;
}

export interface PartyMember {
  userId: string;
  username: string;
  avatarUrl?: string;
  role: 'leader' | 'member';
  joinedAt: string;
  // Party UX fields — data layer always supplies these.
  user: SocialUser;
  isReady: boolean;
  isSpeaking: boolean;
}

export interface CommunityPost {
  id: string;
  authorId: string;
  authorUsername: string;
  authorAvatar?: string;
  title?: string;
  content: string;
  category: string;
  likes: number;
  comments: number;
  shares?: number;
  isLiked: boolean;
  isPinned?: boolean;
  createdAt: string;
  tags?: string[];
  media?: PostMedia[];
  author?: {
    id: string;
    username: string;
    avatar?: string;
    elo?: number;
    status?: string;
  };
}

export interface PostMedia {
  id: string;
  url: string;
  type: 'image' | 'video';
  thumbnail?: string;
}

export interface CommunityComment {
  id: string;
  postId: string;
  author: SocialUser;
  content: string;
  likes: number;
  isLiked: boolean;
  createdAt: string;
}

export interface OnlineStatus {
  userId: string;
  username: string;
  isOnline: boolean;
  lastSeen?: string;
  statusMessage?: string;
}

export interface SocialNotification {
  id: string;
  type: 'friend_request' | 'message' | 'party_invite' | 'post_like' | 'post_comment' | 'like';
  title: string;
  message: string;
  fromUser?: SocialUser;
  userId?: string;
  fromUserId?: string;
  fromUsername?: string;
  notificationType?: string;
  content?: string;
  isRead?: boolean;
  read?: boolean;
  relatedId?: string;
  createdAt: string;
}

export interface FriendsListResponse {
  friends: Friend[];
  totalCount: number;
  onlineCount: number;
}

export interface PartyInvite {
  id: string;
  partyId: string;
  partyName: string;
  inviter: SocialUser;
  invitedUser: SocialUser;
  createdAt: string;
  status: 'pending' | 'accepted' | 'declined';
}

export interface SocialStats {
  totalFriends: number;
  onlineFriends: number;
  totalMessages: number;
  partiesJoined: number;
  communityPosts: number;
  totalLikes: number;
  weeklyActivity: {
    messagesSent: number;
    gamesPlayed: number;
    timeOnline: number;
  };
}

File: backend/src/models/social.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Friend {
    pub id: Uuid,
    pub username: String,
    pub avatar_url: Option<String>,
    pub is_online: bool,
    pub last_seen: Option<DateTime<Utc>>,
    pub added_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FriendRequest {
    pub id: Uuid,
    pub from_user_id: Uuid,
    pub from_username: String,
    pub from_avatar: Option<String>,
    pub to_user_id: Uuid,
    pub status: String, // pending, accepted, rejected
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub id: Uuid,
    pub from_user_id: Uuid,
    pub from_username: String,
    pub to_user_id: Uuid,
    pub content: String,
    pub is_read: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Conversation {
    pub id: Uuid,
    pub participant_id: Uuid,
    pub participant_username: String,
    pub participant_avatar: Option<String>,
    pub last_message: Option<String>,
    pub last_message_at: Option<DateTime<Utc>>,
    pub unread_count: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Party {
    pub id: Uuid,
    pub leader_id: Uuid,
    pub leader_username: String,
    pub name: String,
    pub description: Option<String>,
    pub max_members: i32,
    pub current_members: i32,
    pub members: Vec<PartyMember>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PartyMember {
    pub user_id: Uuid,
    pub username: String,
    pub avatar_url: Option<String>,
    pub role: String, // leader, member
    pub joined_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommunityPost {
    pub id: Uuid,
    pub author_id: Uuid,
    pub author_username: String,
    pub author_avatar: Option<String>,
    pub title: String,
    pub content: String,
    pub category: String,
    pub likes: i32,
    pub comments: i32,
    pub is_liked: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OnlineStatus {
    pub user_id: Uuid,
    pub username: String,
    pub is_online: bool,
    pub last_seen: Option<DateTime<Utc>>,
    pub status_message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SocialNotification {
    pub id: Uuid,
    pub user_id: Uuid,
    pub notification_type: String, // friend_request, message, party_invite, post_like, post_comment
    pub from_user_id: Option<Uuid>,
    pub from_username: Option<String>,
    pub content: String,
    pub is_read: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddFriendRequest {
    pub friend_id: Uuid,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendMessageRequest {
    pub to_user_id: Uuid,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreatePartyRequest {
    pub name: String,
    pub description: Option<String>,
    pub max_members: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreatePostRequest {
    pub title: String,
    pub content: String,
    pub category: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FriendsListResponse {
    pub friends: Vec<Friend>,
    pub total_count: i32,
    pub online_count: i32,
}
